src/compiler/checker.ts

```diff
@@ -15568,13 +15714,25 @@ namespace ts {
                 bindSourceFile(file, compilerOptions);
             });
 
+            let mergeAugmentations = false;
             // Initialize global symbol table
             forEach(host.getSourceFiles(), file => {
                 if (!isExternalOrCommonJsModule(file)) {
                     mergeSymbolTable(globals, file.locals);
                 }
+                mergeAugmentations = mergeAugmentations || file.moduleAugmentations.length > 0;
             });
 
+            if (mergeAugmentations) {
+                // merge module augmentations.
+                // this needs to be done after global symbol table is initialized to make sure that all ambient modules are indexed 
+                for (const file of host.getSourceFiles()) {
+                    for (const augmentation of file.moduleAugmentations) {

```

Comment:
Maybe just collect all augmentations into a single array and then `for..of` on that so you don't have to visit all the source files again. Very few source files are going to have augmentations. Code would be shorter too.